src/compiler/checker.ts

```diff
@@ -14186,19 +14235,102 @@ namespace ts {
+                            error(node.name, Diagnostics.Ambient_module_declaration_cannot_specify_relative_module_name);
+                        }
+                    }
+                    else {
+                        if (isGlobalAugmentation) {
+                            error(node.name, Diagnostics.Augmentations_for_the_global_scope_can_only_be_directly_nested_in_external_modules_or_ambient_module_declarations);
+                        }
+                        else {
+                            // Node is not an augmentation and is not located on the script level.
+                            // This means that this is declaration of ambient module that is located in other module or namespace which is prohibited.
+                            error(node.name, Diagnostics.Ambient_modules_cannot_be_nested_in_other_modules_or_namespaces);
+                        }
                     }
                 }
             }
             checkSourceElement(node.body);
         }
 
+        function checkBodyOfModuleAugmentation(node: Node, isGlobalAugmentation: boolean): void {
+            switch (node.kind) {
+                case SyntaxKind.VariableStatement:
+                    // error each individual name in variable statement instead of marking the entire variable statement
+                    for (const decl of (<VariableStatement>node).declarationList.declarations) {
+                        if (isBindingPattern(decl.name)) {
+                            for (const el of (<BindingPattern>decl.name).elements) {
+                                // mark individual names in binding pattern
+                                checkBodyOfModuleAugmentation(el, isGlobalAugmentation);
+                            }
+                        }
+                        else {
+                            checkBodyOfModuleAugmentation(decl, isGlobalAugmentation);
+                        }
+                    }
+                    break;
+                case SyntaxKind.ExportAssignment:
+                case SyntaxKind.ExportDeclaration:
+                    grammarErrorOnFirstToken(node, Diagnostics.Exports_and_export_assignments_are_not_permitted_in_module_augmentations);
+                    break;
+                case SyntaxKind.ImportEqualsDeclaration:
+                    if ((<ImportEqualsDeclaration>node).moduleReference.kind !== SyntaxKind.StringLiteral) {
+                        error((<ImportEqualsDeclaration>node).name, Diagnostics.Module_augmentation_cannot_introduce_new_names_in_the_top_level_scope);
+                        break;
+                    }
+                    // fallthrough
+                case SyntaxKind.ImportDeclaration:
+                    grammarErrorOnFirstToken(node, Diagnostics.Imports_are_not_permitted_in_module_augmentations_Consider_moving_them_to_the_enclosing_external_module);
+                    break;
+                default:
+                    const symbol = getSymbolOfNode(node);

```

Comment:
You're assuming here that `getSymbolOfNode` can be called on *any* possible top-level, and that if that produces a symbol, the node is a declaration node. Not sure this is safe, particularly since you might see statements here (that would be flagged as errors elsewhere). I think the `default:` clause above should be replaced with a list of actual node kinds that you want to check.